### PR TITLE
E2E test for Unhealthy GameServer on process crash

### PR DESCRIPTION
**Pull request by markmandel**
More preparation for #956

### CODE REVIEW

File: test/e2e/gameserver_test.go [hunk 1]
```diff
@@ -211,6 +211,27 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
...
+	if err != nil {
+		t.Fatalf("Could not get a GameServer ready: %v", err)
+	}
+
+	gsClient := framework.AgonesClient.AgonesV1().GameServers(defaultNs)
+	defer gsClient.Delete(readyGs.ObjectMeta.Name, nil) // nolint: errcheck
+
+	// there is no response, so don't wait for the timeout
+	go func() {
+		// ignore the error, we don't care about the response
```
Comment (roberthbailey) [inline test/e2e/gameserver_test.go]: Is there any chance that this call fails? If so this test will become flaky. If you look at the response and retry on errors (until the server acks the crash) then the test will be more reliable.

Comment (markmandel) [reply test/e2e/gameserver_test.go]: Admittedly it's a bit hacky. What I'm doing is ignoring the error that will come back when we don't get a response, because I didn't feel like writing a separate method to just send, and not wait for an ACK response.

While it's super unlikely there will fail (we use this all through e2e with no retries), I will admit to cutting a couple of corners here :confused: 

I'll update the test. I can't look at myself in the mirror anymore out of embarrassment :smile:

Comment (markmandel) [reply test/e2e/gameserver_test.go]: Of course, now I've updated it, it won't pass... how very strange.

**Comment by markmandel** [comment]
I feel like this might be a race condition of some kind. Maybe the health.go file is not receiving the pod shutdown event / the event is being merged into the restart event - which may be possible.

Attaching zips of the udp-simple logs (json & csv format) - can see the container receiving the CRASH message, so that is definitely happening.
[container_cluster_name_e2e-test-cluster_namespace_id_default__logs__2019-09-05T03-50.zip](https://github.com/googleforgames/agones/files/3581643/container_cluster_name_e2e-test-cluster_namespace_id_default__logs__2019-09-05T03-50.zip)

Then we see the container restart, and the binary attempt to re-register as Ready. Hard to see on the e2e cluster if it actually gets to ready or not.

For this test, we could look at the `status.container.lastState` as well as the current state. I'm not sure how that will help solve #956 as that could cause Unhealthy statuses to occur when we don't want them to. But maybe one step at a time.

I'll go hunt down the relevant docs in the game server controller as well.

At least this will help lock down the behaviour when things go wrong in GameServers.

**Comment by markmandel** [comment]
I can reproduce this locally, if I run all of `gameserver_test.go`. 

This is the logs from the health controller for the gameserver in question - as you can see, there is no moving to Unhealthy.

```
{"gsKey":"default/udp-server8qh7b","message":"Enqueuing","queue":"agones.dev.HealthController","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:15.904024795Z"}
{"gsKey":"default/udp-server8qh7b","message":"Processing","queue":"agones.dev.HealthController","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:15.909286119Z"}
{"gsKey":"default/udp-server8qh7b","message":"Synchronising","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:15.909381042Z"}
{"gsKey":"default/udp-server8qh7b","message":"Enqueuing","queue":"agones.dev.HealthController","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:17.719181804Z"}
{"gsKey":"default/udp-server8qh7b","message":"Enqueuing","queue":"agones.dev.HealthController","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:17.72305824Z"}
{"gsKey":"default/udp-server8qh7b","message":"Processing","queue":"agones.dev.HealthController","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:17.724500147Z"}
{"gsKey":"default/udp-server8qh7b","message":"Synchronising","severity":"info","source":"*gameservers.HealthController","time":"2019-09-06T00:17:17.724567851Z"}
```

The gameserver ends up in Ready state, and never moves. So yeah, race condition of some kind!

**Comment by markmandel** [comment]
Okay, test this new solution multiple times - can't get it to fail anymore.

File: test/e2e/gameserver_test.go [hunk 2]
```diff
@@ -211,6 +213,34 @@ func TestGameServerUnhealthyAfterDeletingPod(t *testing.T) {
...
 
+func TestGameServerUnhealthyAfterReadyCrash(t *testing.T) {
+	t.Parallel()
+
+	l := logrus.WithField("test", "TestGameServerUnhealthyAfterReadyCrash")
+
+	gs := defaultGameServer()
+	readyGs, err := framework.CreateGameServerAndWaitUntilReady(defaultNs, gs)
+	if err != nil {
+		t.Fatalf("Could not get a GameServer ready: %v", err)
```
Comment (roberthbailey) [inline test/e2e/gameserver_test.go]: any particular reason this one uses `t.Fatalf(...)` while below you use `assert.NoError(...)`

File: pkg/gameservers/health.go [hunk 3]
```diff
@@ -125,8 +125,11 @@ func (hc *HealthController) unschedulableWithNoFreePorts(pod *corev1.Pod) bool {
 func (hc *HealthController) failedContainer(pod *corev1.Pod) bool {
 	container := pod.Annotations[agonesv1.GameServerContainerAnnotation]
 	for _, cs := range pod.Status.ContainerStatuses {
-		if cs.Name == container && cs.State.Terminated != nil {
-			return true
+		if cs.Name == container {
+			// sometimes on a restart, these cs.State can be running and the last state will be merged
+			if cs.State.Terminated != nil || cs.LastTerminationState.Terminated != nil {
```
Comment (roberthbailey) [inline pkg/gameservers/health.go]: nit: can probably be `return cs.State.Terminated != nil || cs.LastTerminationState.Terminated != nil` to break the loop once the desired container is found.

File: pkg/gameservers/health.go [hunk 4]
```diff
@@ -125,8 +125,11 @@ func (hc *HealthController) unschedulableWithNoFreePorts(pod *corev1.Pod) bool {
 func (hc *HealthController) failedContainer(pod *corev1.Pod) bool {
 	container := pod.Annotations[agonesv1.GameServerContainerAnnotation]
 	for _, cs := range pod.Status.ContainerStatuses {
-		if cs.Name == container && cs.State.Terminated != nil {
-			return true
+		if cs.Name == container {
+			// sometimes on a restart, these cs.State can be running and the last state will be merged
```
Comment (roberthbailey) [inline pkg/gameservers/health.go]: nit: s/these/the/

Comment (roberthbailey) [inline pkg/gameservers/health.go]: question - was this a bug before where there was a race between checking the health and the container restarting?
question2 - it seems like this will always return true if the container has *ever* failed, which is a change from before (where it was transient but could flip back to false after a restart). Is that the correct behavior?

File: test/e2e/gameserver_test.go (hunk 2 above)
Comment (markmandel) [reply test/e2e/gameserver_test.go]: Mostly - because copy paste :smile: it's what we have everywhere..

Lower down, it would be super rare to fail (how could you not write to a buffer that doesn't track if it gets received or not?)

File: pkg/gameservers/health.go (hunk 3 above)
Comment (markmandel) [reply pkg/gameservers/health.go]: Yes! Change made.

File: pkg/gameservers/health.go (hunk 4 above)
Comment (markmandel) [reply pkg/gameservers/health.go]: These are good questions:

question1 - So the bug was a couple of things:
1) The issue outlined in #956 , but also, that could potentially not happen (i.e. a race condition could mean on restart, it would remain in a Ready state) -- so it was inconsistent. At least now, it consistently moves to an unhealthy state on crash.

2) This is essentially true. If the pod has ever failed (in the last 2 iterations, because that is all that is tracked), and the pod gets updated (say the sidecar crashes), we will move to unhealthy - but at least consistently so.

The next PR I'm going to do, is actually solve #956. My current thinking is - when a GameServer goes to Ready, we put an annotation on the pod with the current `Running` gameserver container ID. Then we always know what not to look at before - i.e. if there was a status.State of Running, and it matched our annotation, we would know not to look at the LastTerminatedState, since that came before.

If that annotation is not there, we don't ever move to Unhealthy, and allow restarts to occur as per normal K8s.

There might be some other edge cases in there -- but I think that gets us to what is the currently documented health checking strategy.

TL;DR - this is an interim step.

Comment (markmandel) [reply pkg/gameservers/health.go]: I just realised that 1.0 RC comes out Tuesday. Happy for this to be merged post 1.0 stable release. It means we have #956 outstanding for 1.0, but I don't think it's a hard blocker for 1.0.

I wouldn't want this to be included in 1.0 without the follow up PR to fix #956 - I've started work on it, but probably safer to wait.

**Comment by roberthbailey** [comment]
When I rebased to head, this PR lost the changes moving the simple udp container version from 0.14 to 0.15 since that version bump was done as part of https://github.com/googleforgames/agones/pull/1059. So we need to move from 0.15 to 0.16 in this PR now.

/hold

**Comment by markmandel** [comment]
Did a local rebase - I think this should be good to go now @roberthbailey

**Comment by roberthbailey** [comment]
Do we not need to bump the simple-udp container version any longer?

**Comment by markmandel** [comment]
There are no changes to the simple-udp implementation, so I'm not sure why we need to bump it?

**Comment by markmandel** [comment]
Oh yeah, I wrote in the description for this PR that there was a change in udp-simple. But i don't see any. Weird. I forget when happened? Whomever merges/squashes this, please feel free to remove that line from the commit.

**Comment by roberthbailey** [comment]
Is there a way to modify commit messages when merging a PR?

**Comment by markmandel** [comment]
Yeah, when you click "merge" it squashes it and gives you the option to edit the commit message on GitHub.

**Comment by roberthbailey** [comment]
Neat. I guess I'd always clicked through that part quickly without reading it before.
